### PR TITLE
Improved kobayashi model and example

### LINKED ISSUE #6463: The anisotropic interface kernels and examples need to be improved
### Description of the enhancement or error report
The ACInterfaceKobayashi kernels have errors in the jacobian. Also, the snow examples have a mesh that is too coarse, severely impacting the solution.

### Rationale for the enhancement or information for reproducing the error
The incorrect Jacobian hurts convergence such that very small time steps are needed. The coarse mesh induces branching independent of the anisotropy.

### Identified impact
This model is our first anisotropic interface model and needs to be as correct as possible, as it will serve as an example for future development in this area.

### PR DESCRIPTION
**Pull request by tonkmr**
I made fairly substantial changes to ACInterfaceKobayashi1 and ACInterfaceKobayashi2. The largest design change is I moved all things related to epsilon into the InterfaceOrientationMaterial. So, a user can change the anisotropic interfacial parameter without changing the kernels. I also cleaned things up so it is easier to read.

Unfortunately, I think there is still an error in the Jacobian. It solves very well with PJFNK but not as well with Newton. The JacobianDebugger indicates it is an on-diagonal term for the order parameter, but I haven't been able to find it.

Finally, in the example and the test, I moved the T dependence on the free energy into the free energy material rather than an auxvariable. So, now T and w are tightly coupled while they were only loosely coupled previously.

Note, that I also added functionality to RankTwoTensor. I thought I would need it for this, but I ended up not using it. However, it is important capability that should be there. I added the capability for RankTwoTensors to be multiplied by a RealVectorValue and return a RealVectorValue.

Closes #6463

### CODE REVIEW

**Comment by dschwen** [comment]
@tonkmr how does the new model perform compared with the old one? Did you get it to solve faster / with larger timesteps?

File: modules/phase_field/src/kernels/ACInterfaceKobayashi1.C [hunk 1]
```diff
@@ -39,69 +42,39 @@ ACInterfaceKobayashi1::ACInterfaceKobayashi1(const InputParameters & parameters)
...
 RealGradient
 ACInterfaceKobayashi1::precomputeQpResidual()
 {
-  // Set interfacial part of residual
-  RealVectorValue v = _eps[_qp]* _deps[_qp] * _L[_qp] * _grad_u[_qp];
-  std::swap(v(0), v(1));
-  v(0) *= -1.0;
-  return v;
+  // Set modified gradient vector
+  RealGradient v( - _grad_u[_qp](1), _grad_u[_qp](0), 0 );
```
Comment (dschwen) [inline modules/phase_field/src/kernels/ACInterfaceKobayashi1.C]: We don't put spaces on the interior of parens. And I would suggest making it `const`.

File: modules/tensor_mechanics/src/utils/RankTwoTensor.C [hunk 2]
```diff
@@ -122,6 +122,14 @@ RankTwoTensor::zero()
 }
 
 void
+RankTwoTensor::vectorOuterProduct(const TypeVector<Real> & a, const TypeVector<Real> & b)
+{
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+      for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/utils/RankTwoTensor.C]: indentation

File: modules/tensor_mechanics/src/utils/RankTwoTensor.C [hunk 3]
```diff
@@ -358,6 +366,19 @@ RankTwoTensor::operator/(const Real b) const
...
 
+TypeVector<Real>
+RankTwoTensor::operator*(const TypeVector<Real> & b) const
+{
+  RealVectorValue result;
+  const RankTwoTensor &a = *this;
+
+  for (unsigned int i = 0; i < N; ++i)
+    for (unsigned int j = 0; j < N; ++j)
+        result(i) += a(i,j) * b(j);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/utils/RankTwoTensor.C]: indentation

File: modules/phase_field/src/materials/InterfaceOrientationMaterial.C [hunk 4]
```diff
@@ -32,15 +42,37 @@ void
...
+  Real d2eps = - _eps_bar * _delta * _delta * _delta * std::cos(_j * (angle - _theta0 * libMesh::pi/180.0));
+
+  // Compute derivative of n with respect to grad_op
+  RealGradient dndgrad_op;
+  if (_grad_op[_qp].norm_sq() == 0)
+    dndgrad_op = 0;
+  else
+  {
+    dndgrad_op(0) = _grad_op[_qp](1) * _grad_op[_qp](1);
+    dndgrad_op(1) = - _grad_op[_qp](0) * _grad_op[_qp](1);
```
Comment (dschwen) [inline modules/phase_field/src/materials/InterfaceOrientationMaterial.C]: Just making sure that the second factor is supposed to have an index 1 in both lines...

**Comment by tonkmr** [comment]
Yes, it does solve faster then before and can take larger time steps. It also produces a cleaner result, probably because I am now fully coupling T and w.

File: modules/phase_field/src/materials/InterfaceOrientationMaterial.C (hunk 4 above)
Comment (tonkmr) [inline modules/phase_field/src/materials/InterfaceOrientationMaterial.C]: Yeah, it is. It comes from n = grad_op(0)/grad_op.norm().

Keep looking for errors, I would love to get the Jacobian perfect.

File: modules/phase_field/examples/anisotropic_interfaces/snow.i [hunk 5]
```diff
@@ -79,62 +67,70 @@
...
+  [../]
+[]
+
 [Executioner]
   type = Transient
-  solve_type = NEWTON
+  scheme = bdf2
+  solve_type = PJFNK
+  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart -pc_hypre_boomeramg_strong_threshold'
+  petsc_options_value = 'hypre    boomeramg      31                 0.7'
```
Comment (permcody) [inline modules/phase_field/examples/anisotropic_interfaces/snow.i]: This is a poor threshold for a 2D problem. MOOSE provides good default boomeramg thresholds for both 2D and 3D problems. I recommend that you don't set this in any input file. You can confirm this by inspecting the header at the top of your simulation output. It'll tell you the threshold being used when using hypre.

**Comment by permcody** [comment]
I see that you submitted a new test and even a gold file but didn't include a "tests" file - SNEAKY!

**Comment by tonkmr** [comment]
Actually, I didn't submit a new test, I just modified the existing test so no changes to the "tests" file was needed.

**Comment by tonkmr** [comment]
This is the first time that I have had someone else commit a new function with the same name that does the same thing as one I have added while waiting for a pull request to be merged.

**Comment by permcody** [comment]
Ah I see, thanks!

File: modules/tensor_mechanics/src/utils/RankTwoTensor.C [hunk 6]
```diff
@@ -369,6 +369,19 @@ RankTwoTensor::operator/(const Real b) const
   return result;
 }
 
+TypeVector<Real>
+RankTwoTensor::operator*(const TypeVector<Real> & b) const
```
Comment (permcody) [inline modules/tensor_mechanics/src/utils/RankTwoTensor.C]: I'm really surprised that this didn't exist until now.

**Comment by permcody** [comment]
This looks good. @dschwen - any further comments?

**Comment by dschwen** [comment]
No, looks all good to me :+1:

File: modules/tensor_mechanics/src/utils/RankTwoTensor.C (hunk 6 above)
Comment (tonkmr) [inline modules/tensor_mechanics/src/utils/RankTwoTensor.C]: Me too.

**Pull request merged**